### PR TITLE
Delete feature

### PR DESCRIPTION
**Pull request by GregDick**
Creating tests for a POST request to (delete/:id)

### CODE REVIEW

**Comment by GregDick** [comment]
@sscotth It's not great but we could use some feedback... Thanks!

**Comment by GregDick** [comment]
@sscotth We merged with Post and all tests are passing now. We would love some feedback

**Comment by GregDick** [comment]
@sscotth Ready for review

File: app/delete_feature/controller.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+'use strict';
+
+var database = require('../../lib/mongo/');
+var Post = require('../post/Post');
+
+module.exports.setHidden = function (req, res){
```
Comment (sscotth) [inline app/delete_feature/controller.js]: move this to the post controller

File: app/routes.js [hunk 2]
```diff
@@ -4,9 +4,12 @@ var express = require('express');
 var router = express.Router();
 
 var home = require('./home/routes');
+var delete_feature = require('./delete_feature/routes');
 var post = require('./post/routes');
 
 router.use('/', home);
 router.use('/', post);
+router.use('/delete', delete_feature);
```
Comment (sscotth) [inline app/routes.js]: No need to create a new root route. Use a DELETE request to `/post/:id`

File: app/index.js [hunk 3]
```diff
@@ -7,6 +7,7 @@ var morgan = require('morgan');
 var sass = require('node-sass-middleware');
 
 var routes = require('./routes');
+
```
Comment (sscotth) [inline app/index.js]: Be more conscious of extra spaces in your commits.

File: app/delete_feature/routes.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+'use strict';
+
+var express = require('express');
+var router = express.Router();
+
+var ctrl = require('./controller');
+
+router.post('/:id', ctrl.setHidden);
```
Comment (sscotth) [inline app/delete_feature/routes.js]: Move this to the post routes and make it a delete request

File: app/delete_feature/controller.js (hunk 1 above)
Comment (sscotth) [inline app/delete_feature/controller.js]: Move this logic to the Post model

**Comment by sscotth** [comment]
You have a good amount of logic as it relates to different users, which we don't have yet. If you remove this logic, then I can merge it. Then, using a new branch/PR you can re-add it.

**Comment by GregDick** [comment]
@sscotth Took out user logic and changed post request to delete request. Ready for review

File: app/post/controller.js [hunk 5]
```diff
@@ -16,6 +16,15 @@ module.exports.create = function (req, res) {
   });
 };
 
+module.exports.setHidden = function (req, res) {
+  Post.findById(req.params.id, function (err, post) {
```
Comment (sscotth) [inline app/post/controller.js:20]: Look and see if there is a mongo method to find and modify in on call to the database

File: app/post/controller.js [hunk 6]
```diff
@@ -16,6 +16,15 @@ module.exports.create = function (req, res) {
...
+    Post.setHidden(post, function (err, writeResult){
+      if (err) { throw err; }
+      res.send(writeResult);
```
Comment (sscotth) [inline app/post/controller.js]: Something needs to happen when a delete action is fired, and sending raw data to the user is certainly not good user experience. Consider a redirect and maybe a flash message.

File: app/post/Post.js [hunk 7]
```diff
@@ -23,6 +23,10 @@ Post.create = function (post, cb) {
   Post.collection.insertOne(post, cb);
 };
 
+Post.setHidden = function (post, cb){
```
Comment (sscotth) [inline app/post/Post.js:26]: Can you write a unit test for this function in the `Post.test.js` file?

File: app/delete_feature/delete.test.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    });
+  });
+
+  //drop collections after tests
+  after(function (done){
+      Post.dropCollection(done);
+    });
+
+  //  test route
+  it('should respond with success', function (done){
```
Comment (sscotth) [inline app/delete_feature/delete.test.js]: remove this test as it is being tested more fully below.

File: app/delete_feature/delete.test.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+      });
+
+  //  adding hidden to a post
+  it('should add hidden to the post', function (done){
+      request(app)
+        .delete('/post/' + seededPost[0]._id)
+        .expect(200)
+        .end(function (err){
+          if (err) throw err;
+          Post.findById(seededPost[0]._id, function (err, post){
```
Comment (sscotth) [inline app/delete_feature/delete.test.js]: For this acceptance test, or one that actually tests HTTP responses, use the output of the html. Do not connect to the database. Database calls are used for integration tests. So move this logic to a test strictly of the `Post.findById` function.

Preferred: send a delete request, and then make another request to see if the data is visible on the page.

**Comment by GregDick** [comment]
@sscotth Ready for review!

**Comment by sscotth** [comment]
Almost there. Can you move your acceptance test that is currently in `delete.test.js` to the `Post/routes.test.js` file? Also, can you move the test I mention below to the root "index" route? Also add a preliminary check that the text is there.

So the test would look like this
request /, res.text should contain, delete :id, request /, res.text should not contain

Finally, clean up any extra spaces and comments that you have left behind. You can view the changes that you have made in the "Files Changed" tab above.

File: app/delete_feature/delete.test.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        .delete('/post/' + seededPost[0]._id)
+        .expect(302)
+        .end(function (err){
+          if (err) throw err;
+          request(app)
+            .get('/post/' + seededPost[0]._id)
+            .expect(302)
+            .end(function (err){
+              if (err) throw err;
+              request(app)
```
Comment (sscotth) [inline app/delete_feature/delete.test.js]: Move this request into its own `it` block

File: app/delete_feature/delete.test.js [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+  //  adding hidden to a post
+  it('should add hidden to the post', function (done){
+      request(app)
+        .delete('/post/' + seededPost[0]._id)
+        .expect(302)
+        .end(function (err){
+          if (err) throw err;
+          request(app)
+            .get('/post/' + seededPost[0]._id)
+            .expect(302)
```
Comment (sscotth) [inline app/delete_feature/delete.test.js]: Should a deleted resource return 302, or should it 404?

File: app/delete_feature/delete.test.js [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+  //  adding hidden to a post
+  it('should add hidden to the post', function (done){
+      request(app)
+        .delete('/post/' + seededPost[0]._id)
+        .expect(302)
+        .end(function (err){
+          if (err) throw err;
+          request(app)
+            .get('/post/' + seededPost[0]._id)
```
Comment (sscotth) [inline app/delete_feature/delete.test.js]: Can you also perform this get request before the delete, to test that it returns properly

**Comment by GregDick** [comment]
@sscotth Refactored tests and cleaned up line spacing. Ready for review

File: app/post/routes.test.js [hunk 13]
```diff
@@ -10,7 +10,7 @@ var Post = require('./Post');
 describe('Post Routes', function () {
   var seededPosts;
 
-  before(function (done) {
+  beforeEach(function (done) {
```
Comment (sscotth) [inline app/post/routes.test.js]: You only need to connect to mongo once, so keep this as `before`

File: app/post/routes.test.js [hunk 14]
```diff
@@ -24,7 +24,7 @@ describe('Post Routes', function () {
     });
   });
 
-  after(function (done) {
+  afterEach(function (done) {
```
Comment (sscotth) [inline app/post/routes.test.js]: You only need to cleanup once, so keep this as `after`

**Comment by GregDick** [comment]
@sscotth changed order of tests to avoid beforeEach and afterEach

File: app/post/controller.js [hunk 15]
```diff
@@ -16,8 +16,22 @@ module.exports.create = function (req, res) {
...
+      if (err) { throw err; }
+      res.redirect('/');
+    });
+  });
+};
+
 module.exports.show = function (req, res) {
   Post.findById(req.params.id, function (err, post) {
-    res.render('post/show', {post: post});
+    if(post.hidden){
```
Comment (sscotth) [inline app/post/controller.js]: Try this:

``` js
module.exports.show = function (req, res, next) {
  Post.findById(req.params.id, function (err, post) {
    if (post.hidden) {
      var err = new Error('Not Found');
      err.status = 404;
      next(err);
    }
```

**Comment by GregDick** [comment]
@sscotth added error handling function

**Comment by sscotth** [comment]
Merged, but I made some changes in 0f23943f508c9b390bda9884dbeb63ba2900d54c let me know if you have any questions

**Pull request merged**